http: use `autoDestroy: true` in incoming message

dnlup opened the pull request:
Enable the default `autoDestroy: true` option in IncomingMessage.

Refactor `_http_client` and `_http_server` to remove any manual destroying/closing of IncomingMessage.
Refactor IncomingMessage `destroy` method to use the standard implementation of the stream module and move the abort logic 
 inside of it.


Ref:

#30625

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)


<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

ronag commented:
`IncomingMessage` overrides `destroy()` which kind of blocks any effort in this regards. That needs to be fixed first.

dnlup commented:
> `IncomingMessage` overrides `destroy()` which kind of blocks any effort in this regards. That needs to IMO be fixed first.

Thanks, @ronag, for the suggestion. I'll look into it. There is an issue with `destroy` for sure. When using keep-alive, we don't want to destroy the socket, and that's what's happening by just setting the option to `true`.

dnlup commented:
Sorry if this is taking a long time, but I am trying to figure out what's wrong :pray: 

Removing the override of `destroy()` resolves the `EECONRESET` errors, fewer tests are failing, but `request` is emitting `close` twice, and some connections seem to hang.

dnlup committed:
http: use `autoDestroy: true` in incoming message

Enable the default `autoDestroy: true` option in IncomingMessage.

Refactor `_http_client` and `_http_server` to remove any manual
destroying/closing of IncomingMessage.
Refactor IncomingMessage `destroy` method to use the standard
implementation of the stream module and move the early termination
event emitting inside of it.

dnlup commented:
The changes made here are overriding the ones you made recently, @ronag . I don't know if this approach is the best.

BridgeAR commented:
@ronag PTAL

ronag commented on lib/_http_incoming.js:
```diff
@@ -114,19 +114,25 @@ IncomingMessage.prototype._read = function _read(n) {
+    // emitCloseNT on the Stream.Readable side because the event won't be
+    // emitted from the server.
+    cb(err);
+  }
```
callback must always be invoked or you get corrupted state, there is not much point in autodestroy if we don't implement it properly

ronag reviewed (changes requested):
Please also try to remove all cases of manually managing `destroyed` and emitting `'close'` on `IncomingMessage`. Having some kind of hybrid solution will make things hard to maintain IMO.

dnlup commented:
> Please also try to remove all cases of manually managing `destroyed` and emitting `'close'` on `IncomingMessage`. Having some kind of hybrid solution will make things hard to maintain IMO.

You mean like in [here](https://github.com/nodejs/node/blob/master/lib/_http_server.js#L675)? I agree, that is also the reason why the callback is not called in all cases in `_destroy`. I also agree that it is ugly, sorry. I wasn't sure I could change the `_http_server.js` parts.

Thanks for the pointers.

dnlup commented:
Do we need to keep [this behavior](https://github.com/nodejs/node/blob/master/lib/_http_server.js#L553)? If we do, I think it would be better to implement the destroy logic directly in `IncomingMessage` and not reusing the general `lib/internal/streams/destroy.js` implementation as I did. Overriding that implementation causes a lot of tests to fail because they are expecting an unhandled error. Also, do you have anything against adding an `abort` method on `IncomingMessage`?

ronag commented:
> Do we need to keep this behavior? 

Yes, just call destroy and move the logic there? 

> If we do, I think it would be better to implement the destroy logic directly in IncomingMessage and not reusing the general lib/internal/streams/destroy.js implementation as I did. 

I disagree. We should implement it as a regular destroy.

> Overriding that implementation causes a lot of tests to fail because they are expecting an unhandled error. 

Can still be an unhandeld error?

> Also, do you have anything against adding an abort method on IncomingMessage?

Yes, destroy should be abort. We just deprecated abort on ClientRequest in favor of destroy.

dnlup commented:
> Yes, just call destroy and move the logic there?

Ok.

> I disagree. We should implement it as a regular destroy.

I agree. The only difference I have seen with a regular destroy is that IncomingMessage emits an error only if there are listeners attached, at least when aborting.

> Yes, destroy should be abort. We just deprecated abort on ClientRequest in favor of destroy.

Got it.

ronag commented:
> The only difference I have seen with a regular destroy is that IncomingMessage emits an error only if there are listeners attached, at least when aborting.

I can help with sorting this out. If you have any test that depends on this behavior just comment them out and we can take a look together when that's the only thing remaining.

dnlup commented on lib/_http_incoming.js:
```diff
@@ -114,19 +114,21 @@ IncomingMessage.prototype._read = function _read(n) {
-  return this;
+IncomingMessage.prototype._destroy = function _destroy(err, cb) {
+  // TODO (dnlup): move abort logic here based on readableEnded state
+  if (!this.readableEnded || !this.complete) {
```
I tried to move the abort logic here. To make it work for both client and server I also had to check for the `readableEnded` state.

dnlup commented on lib/_http_client.js:
```diff
@@ -723,12 +710,14 @@ function requestOnPrefinish() {
   if (req.res) {
-    req.res.emit('close');
+    // TODO(dnlup): is this still needed since using autoDestroy: true?
+    req.res.destroy();
```
I need to to some further testing to see if this is needed after setting `autoDestroy: true`.

dnlup commented on test/parallel/test-http-aborted.js:
```diff
@@ -35,27 +35,27 @@ const assert = require('assert');
-    }));
-  }));
-}
+// {
```
The changes in this PR break this behavior, on both client and server. I am not sure about which direction I should go.

dnlup commented on lib/_http_incoming.js:
```diff
@@ -114,19 +114,21 @@ IncomingMessage.prototype._read = function _read(n) {
-  if (this.socket)
-    this.socket.destroy(error);
-  return this;
+IncomingMessage.prototype._destroy = function _destroy(err, cb) {
```
I was wondering if using the standard destroy implementation could cause some performance regressions since it has more checks than the logic that was in place before.

dnlup replied:
Is there any specific benchmark you would like me to run to check this?

ronag replied:
I don't think we need to worry about destroy performance.

dnlup commented:
I think I was able to exclude all the tests depending on the breaking behavior on the `error` event. I moved the emitting of the `aborted` event inside `_destroy`.

ronag commented:
@dnlup Good work! I'm a bit swamped this week but I hope to help out here next week.

dnlup commented:
Rebased against the latest changes. Still  a bit unsure of which direction to take:

1. intercept the `error` to keep the old behavior (which is emitting the event only if there are listeners attached)
2. refactor tests because we expect this breaking change

dnlup commented on lib/_http_incoming.js:
```diff
@@ -80,6 +82,9 @@ function IncomingMessage(socket) {
   this._dumped = false;
+
+  // Catch errors to avoid crashes now that autoDestroy is enabled
+  this.on('error', noop);
```
Regarding https://github.com/nodejs/node/pull/33035#issuecomment-641124262 I went for option 1

dnlup replied:
This causes this test to fail though:

![image](https://user-images.githubusercontent.com/15520377/88365940-e1281180-cd87-11ea-9434-402fbd9248f2.png)

Is this acceptable? If not, I think I should write a custom `destroy` implementation because the standard one *always* emits an `error` event in case of errors.

dnlup replied:
By acceptable I mean that this behavior is acceptable and we can remove/refactor this test case.

ronag replied:
Sorry but this is a nono. We can't swallow errors like that.

dnlup replied:
Understood. I'll work on a destroy method that implements the logic that was on _http_server.js. Thanks.

dnlup commented:
Regarding https://github.com/nodejs/node/pull/33035#issuecomment-652818333, a third option would be to implement a custom `destroy` method, which emits an `error` event *only* if there are listeners attached. The standard implementation I used always emits an `error` event, and I don't see a way around that.

Any thoughts?

dnlup commented on lib/_http_incoming.js:
```diff
@@ -114,19 +114,20 @@ IncomingMessage.prototype._read = function _read(n) {
+  if (this.socket && !this.readableEnded) {
+    this.socket.destroy(err);
+  }
+  this.listenerCount('error') > 0 ? cb(err) : cb();
```
With this approach, we can emulate a backward-compatible behavior. To be fair though, this does not behave exactly like [before](https://github.com/nodejs/node/blob/master/lib/_http_server.js#L552) because this check was done only when terminating early, while now it is done every time `_destroy` is called with an error. It would be an easy change to move this check a few lines above, if necessary.

ronag replied:
Why not just do this inside !this.readableEnded?

dnlup commented:
Sorry to bother again, this is just a ping to understand if the work done here goes anywhere near the right direction :wink:

ronag reviewed (commented):
This is looking pretty good. Is it ready?

dnlup commented:
> This is looking pretty good. Is it ready?

@ronag I have just [this doubt](https://github.com/nodejs/node/pull/33035/files#r460666733). The job failing on macOS is brand new :smile: , I don't have a machine to verify if the failure is temporary. Other than that, I think it's ready.

ronag commented:
This changes timings slightly so I would consider it semver-major.

ronag commented on lib/_http_incoming.js:
```diff
@@ -160,19 +160,20 @@ IncomingMessage.prototype._read = function _read(n) {
+    this.emit('aborted');
+  }
+  if (this.socket && !this.readableEnded) {
+    this.socket.destroy(err);
```
socket should only be destroyed when aborted? I think this needs a comment.

ronag commented on lib/_http_incoming.js:
```diff
@@ -160,19 +160,20 @@ IncomingMessage.prototype._read = function _read(n) {
+  }
+  this.listenerCount('error') > 0 ? cb(err) : cb();
 };
 
```
I would probably do `_destroy` like this:

```js
if (!this.readableEnded || !this.complete) {
  this.aborted = true;
  this.emit('aborted');
  err = connResetException('aborted');
}

const callback = (er) => {
  cb(this.aborted && this.listenerCount('error') === 0 ? null : (er || err));
};

if (err && this.socket) {
  this.socket.destroy(err);
  finished(this.socket, callback);
} else {
  callback(err);
}

```

dnlup replied:
`finished` would be https://nodejs.org/api/stream.html#stream_stream_finished_stream_options_callback?

I understand your point. Why do we have to wait for the socket to finish before calling `cb`? Is that related to the timing issue you talked about previously? I am just asking because I don't know.

ronag replied:
> understand your point. Why do we have to wait for the socket to finish before calling cb? Is that related to the timing issue you talked about previously? I am just asking because I don't know.

No, it's not necessary per se. Just best practice. You can also skip waiting it is just less "correct".

dnlup replied:
Ok, thank you so much for the clarification

dnlup replied:
Unfortunately, that implementation is causing some problems in tests. The pain points are 2:

1. tests that call `destroy()` without an error now have an `error` event emitted. For example, `test/parallel/test-async-wrap-tlssocket-asyncreset.js`

![Screenshot from 2020-11-19 07-08-20](https://user-images.githubusercontent.com/15520377/99628653-ded47c00-2a36-11eb-8489-d999e69cdcf0.png)

others are erroring with a timeout:

![Screenshot from 2020-11-19 07-17-11](https://user-images.githubusercontent.com/15520377/99628847-47235d80-2a37-11eb-8726-77eb8a83f2f7.png)

2. for some reason I am trying to figure out, it seems the `Stream.finished` callback is not getting called in the client in the test `parallel/test-http-client-spurious-aborted.js` causing the error not to be propagated:

![Screenshot from 2020-11-19 06-34-18](https://user-images.githubusercontent.com/15520377/99629147-c6b12c80-2a37-11eb-8890-3bd287eec558.png)

This implementation has just the problem `2`:

```js
IncomingMessage.prototype._destroy = function _destroy(err, cb) {
  if (!this.readableEnded || !this.complete) {
    this.aborted = true;
    this.emit('aborted');
  }
  if (this.socket && this.aborted) {
    this.socket.destroy(err);
    const cleanup = Stream.finished(this.socket, e => {
      cleanup()
      this.listenerCount('error') > 0 ? cb(e || err) : cb();
    })
  } else {
    this.listenerCount('error') > 0 ? cb(err) : cb();
  }
};
```
Please overlook the code repetition for now. I'll fix it once I find out the problem.

dnlup replied:
I am debugging `test/parallel/test-http-client-spurious-aborted.js`.

It looks like [this check](https://github.com/nodejs/node/blob/master/lib/internal/streams/end-of-stream.js#L163) is always `false` on the client when running the test. On the server-side, that check returns `true` and `callback` is called on the next tick, while on the client-side, that doesn't happen.

dnlup replied:
I tried to implement it as close as possible with the caveats explained earlier. I am now destroying the socket only if it hasn't already been destroyed. That avoids calling finished on a finished stream.

ptal @ronag

dnlup committed:
http: reafactor incoming message destroy

Destroy the underlying socket only if it is not ready destroyed. Wait
for the stream to finish in that case.

dnlup committed:
http: fix lint error in incoming message

ronag commented on lib/_http_incoming.js:
```diff
@@ -160,18 +160,27 @@ IncomingMessage.prototype._read = function _read(n) {
 
+  // If aborted and the underlying socket not already destroyed,
+  // destroy it.
+  if (this.socket && !this.socket.destroyed && this.aborted) {
```
I don't think you need to check socket.destroyed

dnlup replied:
I agree, but I am having problems with this test:

parallel/test-http-client-spurious-aborted.js

Where it seems that the underlying socket is already destroyed when calling `IncomingMessage.destroy`. This causes `finished` to never call the callback.

ronag replied:
ah, `finished` should be able to handle that but maybe the `willEmitClose` check needs improvement.

ronag commented on lib/_http_incoming.js:
```diff
@@ -349,6 +358,14 @@ IncomingMessage.prototype._dump = function _dump() {
   }
 };
 
+function onError(instance, cb, error) {
```
instance => self

ronag commented on lib/_http_incoming.js:
```diff
@@ -349,6 +358,14 @@ IncomingMessage.prototype._dump = function _dump() {
 };
 
+function onError(instance, cb, error) {
+  if (instance.listenerCount('error') > 0) {
```
Maybe a comment describing why we do this?

dnlup replied:
Yup, thanks

ronag commented:
@dnlup How are we doing with tests?

dnlup commented:
@ronag  All tests are passing, the behavior should not be different from before (besides the timing difference you mentioned). The thing that was breaking everything was the fact that an `error` event is emitted only if there's a listener, and my first implementation didn't consider that. I'll add a comment where you said in the review to explain this.

dnlup commented:
Basically, I didn't have to change a single test

ronag reviewed (approved):
Impressive work!

ronag commented:
There is a slight timing difference but otherwise behaves the same. I would like to consider landing this on v15 (not 14 though). @nodejs/http @nodejs/tsc wdyt?

ronag commented on lib/_http_incoming.js:
```diff
@@ -160,18 +160,27 @@ IncomingMessage.prototype._read = function _read(n) {
+  // If aborted and the underlying socket not already destroyed,
+  // destroy it.
+  if (this.socket && !this.socket.destroyed && this.aborted) {
+    this.socket.destroy(err);
```
@dnlup Question. What if `destroy(err || connResetException('aborted'))`? Then maybe we can remove `connResetException('aborted')` from the other 2 places.

dnlup replied:
The problem I encountered with that approach is that some tests, like:

`test/parallel/test-async-wrap-tlssocket-asyncreset.js`

Are calling `destroy` without an error and don't expect an `error` event. They register a handler but check explicitly that it is not called. I interpreted this as "don't emit an error if none was passed".

ronag replied:
Ok. Let's leave it for now. Maybe a future semver-major.

dnlup committed:
http: add comments in _http_incoming

mcollina reviewed (approved):
lgtm, I agree

Trott commented on lib/_http_client.js:
```diff
@@ -752,13 +739,10 @@ function requestOnPrefinish() {
   if (req.socket) {
     req.socket.emit('free');
   }
+
```
Extraneous whitespace change should ideally be removed but I'm not going to block on that!

dnlup replied:
Sorry, I'll fix it. Thanks.

wa-Nadoo commented on lib/_http_incoming.js:
```diff
@@ -349,6 +361,17 @@ IncomingMessage.prototype._dump = function _dump() {
+  // This is to keep backward compatible behavior.
+  // An error is emitted only if there are listeners attached to
+  // the event.
+  if (instance.listenerCount('error') > 0) {
+    cb(error);
+  } else {
+    cb();
+  }
```
IMO this condition is too permissive. In the previous implementation, IncomingMessage doesn't propagate only abort errors if error listener count === 0. After latest changes all errors don't propagated if error listener count === 0.
```suggestion
  if (instance.listenerCount('error') === 0 && error.code === 'ECONNRESET' && error.message === 'aborted') {
    cb();
  } else {
    cb(error);
  }
```

dnlup replied:
Good catch, thanks, @wa-Nadoo . I intended to keep the same behavior, so I should implement the stricter check that you suggested.

ronag replied:
@wa-Nadoo is correct.

ronag replied:
`IncomingMessage.destroy(new Error())` should be unhandled exception

dnlup replied:
I think I got confused, the check is indeed in functions that should be called only when aborting if I am not mistaken. I'll fix this and add a test.

dnlup replied:
I fixed this. Again, thanks @wa-Nadoo and @ronag

dnlup replied:
Sorry, I need to reopen this. I don't think that the behavior was as you explained in the previous implementation. In fact, in `http_server` and `_http_client`,  the points where an error is emitted are in the abort logic, and only if there are listeners attached to the event. That, and the fact the IncomingMessage wasn't implementing a standard destroy (therefore, it wasn't itself emitting an `error` event), made it emit an error event only in case of abort if there were event handlers.

Something that I think proves this is the fact that in `_http_client`, the `error` event of `res` is no bubbled up to `req`. The same goes for `_http_server`. I think maybe this example explains this: https://github.com/nodejs/node/pull/33035#discussion_r543109670.`

I added some tests, ptal.

wa-Nadoo commented on lib/_http_incoming.js:
```diff
@@ -349,6 +361,17 @@ IncomingMessage.prototype._dump = function _dump() {
   }
 };
 
+function onError(instance, cb, error) {
```
Argument order like this is a bit non-standard for the Node.js ecosystem.
```suggestion
function onError(instance, error, cb) {
```

dnlup replied:
Thank you, I'll change it.

dnlup committed:
http: remove trailing space

dnlup committed:
http: use standard args order in IncomingMEssage onError

dnlup commented on test/message/http_client_incoming_message_destroy.js:
```diff
@@ -0,0 +1,16 @@
+'use strict';
```
I opted for testing that the process crashes, are there any alternatives?

ronag replied:
Process.on uncaught

dnlup replied:
Thanks.

dnlup replied:
Sorry for being annoying. While I was working on using `process.on()` in tests, I saw that the behavior in case of error is not the same as it is in version `15.4`. 

For example, this code:

```js
'use strict';

const common = require('../common');
const { createServer, get } = require('http');

const server = createServer(common.mustCall((req, res) => {
  res.writeHead(200);
  res.write('Part of res.');
}));

function onUncaught(error) {
  console.log('process', error)
}

process.on('uncaughtException', onUncaught)

server.listen(0, () => {
  get({
    port: server.address().port
  }, common.mustCall((res) => {
    res.socket.on('error', e => console.log('socket', e))
    res.destroy(new Error('Destroy test'));
  })).on('error', e => console.log('request', e));
});

```

on node `15.4`

gives this output:

```bash
request Error: Destroy test
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/parallel/test-http-client-incomingmessage-destroy.js:23:17)
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/common/index.js:377:15)
    at Object.onceWrapper (node:events:483:26)
    at ClientRequest.emit (node:events:376:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:660:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:126:17)
    at Socket.socketOnData (node:_http_client:526:22)
    at Socket.emit (node:events:376:20)
    at addChunk (node:internal/streams/readable:305:12)
    at readableAddChunk (node:internal/streams/readable:280:9)
socket Error: Destroy test
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/parallel/test-http-client-incomingmessage-destroy.js:23:17)
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/common/index.js:377:15)
    at Object.onceWrapper (node:events:483:26)
    at ClientRequest.emit (node:events:376:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:660:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:126:17)
    at Socket.socketOnData (node:_http_client:526:22)
    at Socket.emit (node:events:376:20)
    at addChunk (node:internal/streams/readable:305:12)
    at readableAddChunk (node:internal/streams/readable:280:9)
```

While on this branch:

```bash
request Error: Destroy test
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/parallel/test-http-client-incomingmessage-destroy.js:23:17)
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/common/index.js:377:15)
    at Object.onceWrapper (node:events:483:26)
    at ClientRequest.emit (node:events:376:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:648:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:126:17)
    at Socket.socketOnData (node:_http_client:514:22)
    at Socket.emit (node:events:376:20)
    at addChunk (node:internal/streams/readable:305:12)
    at readableAddChunk (node:internal/streams/readable:280:9)
socket Error: Destroy test
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/parallel/test-http-client-incomingmessage-destroy.js:23:17)
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/common/index.js:377:15)
    at Object.onceWrapper (node:events:483:26)
    at ClientRequest.emit (node:events:376:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:648:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:126:17)
    at Socket.socketOnData (node:_http_client:514:22)
    at Socket.emit (node:events:376:20)
    at addChunk (node:internal/streams/readable:305:12)
    at readableAddChunk (node:internal/streams/readable:280:9)
process Error: Destroy test
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/parallel/test-http-client-incomingmessage-destroy.js:23:17)
    at ClientRequest.<anonymous> (/home/dnlup/Dev/dnlup/node/test/common/index.js:377:15)
    at Object.onceWrapper (node:events:483:26)
    at ClientRequest.emit (node:events:376:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:648:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:126:17)
    at Socket.socketOnData (node:_http_client:514:22)
    at Socket.emit (node:events:376:20)
    at addChunk (node:internal/streams/readable:305:12)
    at readableAddChunk (node:internal/streams/readable:280:9)
```

It causes an `uncaughtException` even with a `error` handler attached to the `req` object. This is because now `res` is emitting an `error` and before it wasn't, I think. If we intend to keep this behavior, we should handle this in the `_http_client` (listen to `res.emit('error')` and emit an bubble up the event).

I  apologize if this statements are obvious to you, I am just trying to avoid big changes from the current implementation.

dnlup committed:
http: add test for incomingmessage destroy

Test uncaught exceptions when destroying IncomingMessage.

dnlup commented:
I realized I need to fix the error that is emitted when calling `destroy`. Before these changes we emitted an `abort` error in all cases, now we don't.

ronag commented:
@dnlup Excellent work! I am so happy we have fixed this!

dnlup commented:
> @dnlup Excellent work! I am so happy we have fixed this!

@ronag Me too! Thank you for all the help, thank you everyone!

kanongil commented:
This change breaks all request processing in hapi (and probably other uses as well) by emitting `'close'` on the incoming request before the response has even been sent.

This is a semver major change, as the `http.IncomingMessage` `'close'` event is explicitly documented to be emitted once the underlying connection has been closed, and it still shares the connection with the `http.ServerResponse`.

ronag commented:
The documentation is wrong as the socket can be re-used, which was true even before this PR. However, I do agree that emitting `'close'` on the req before res has been ended is a breaking change albeit a correct behavior.

ronag commented:
@targos: It might be worth to consider reverting this on v15?

@dnlup: Would you be interested in opening another PR to add a regression test for req/res 'close' event ordering, i.e. the behavior introduced in this PR.

dnlup commented:
> @dnlup: Would you be interested in opening another PR to add a regression test for req/res 'close' event ordering? i.e. the behavior introduced in this PR.

Yes, of course

kanongil commented:
I agree that this is more correct, but it breaks the existing expectation.

FYI, the request `'close'` is only emitted when the `req` data is consumed, even when empty. Ie. it will still work correctly for many `GET` requests, where the server logic does not consume the data (`req.on('data', …)`).

dnlup commented:
I was thinking about using this test suite here:

test/parallel/test-http-req-res-close.js

@kanongil @ronag By running this test with version 15.5, I should expect `req` emitting close before `res`, am I understanding correctly?

kanongil commented:
That test won't cover this case, since the response is immediately ended, and it does not read from the request.

For inspiration I made this code, which will incorrectly output `REQ CLOSED` on 15.5.0:

```js
const Http = require('http');

const server = Http.createServer((req, res) => {

    console.log('REQ', req.method, req.url)

    req.on('data', (chunk) => {

        console.log('CHUNK', chunk.length)
    });

    req.on('close', () => console.log('REQ CLOSED'));
    res.on('close', () => console.log('RES CLOSED'));
});

server.listen(() => {

    const port = server.address().port;

    const req = Http.request({ port, method: 'GET' }, (res) => {

        console.log('GOT RES', res.statusCode);
        server.close();
    });

    req.end();
});
```

dnlup commented:
> ```js
> const Http = require('http');
> 
> const server = Http.createServer((req, res) => {
> 
>     console.log('REQ', req.method, req.url)
> 
>     req.on('data', (chunk) => {
> 
>         console.log('CHUNK', chunk.length)
>     });
> 
>     req.on('close', () => console.log('REQ CLOSED'));
>     res.on('close', () => console.log('RES CLOSED'));
> });
> 
> server.listen(() => {
> 
>     const port = server.address().port;
> 
>     const req = Http.request({ port, method: 'GET' }, (res) => {
> 
>         console.log('GOT RES', res.statusCode);
>         server.close();
>     });
> 
>     req.end();
> });
> ```

Thank you @kanongil , now I can see the difference.

mcollina commented:
We should revert this in v15 :/.

dnlup commented:
I apologize

targos commented:
Someone please open a revert PR against v15.x-staging

ronag commented:
> I apologize

Don’t worry. These things happen. Nobody’s fault. We’ll sort it out.

ronag commented:
@dnlup do you want to try opening a revert or against v15?

ronag commented:
@nodejs/releasers we should probably add a note to the release notes for v16 regarding the change in behavior. Just ping me and I’ll be happy to author something.

dnlup commented:
> @dnlup do you want to try opening a revert PR against v15?
@ronag Yes, I am going to try. I should revert all the 7 commits in `master` and open a PR to `v15-staging branch`, right?

https://github.com/nodejs/node/compare/eb14b103701650d89e060f61c0e527e43a9e8c39...8154e47e2b407a906f2c1270e5fc76fe466cc886

dnlup commented:
Sorry if it took me a while:

https://github.com/nodejs/node/pull/36647

simhnna commented:
Sorry to revive an ancient discussion, but what how did you intend client aborts to be handled after this change? It looks like currently the only way is to listen on close of responses and check if writableFinished is set to false.

I opened https://github.com/nodejs/node/issues/46666 and more or less completely forgot about it 😅, since it didn't get any attention I [wrote](https://github.com/nodejs/node/pull/64693) something myself, but I'm not sure if the current way should just be documented, or if the abort event should be called again. Maybe you can chime in?
